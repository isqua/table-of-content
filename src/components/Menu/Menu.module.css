.menu {
    font-size: 13px;
    font-style: normal;
    font-weight: 400;
    line-height: 20px;

    --menu-item-block-padding: 8px;
    --menu-item-inline-padding: 22px;
    --menu-item-level-padding: 16px;
}

.list {
    display: flex;
    flex-direction: column;
    padding: 0;
    margin: 0;
    list-style-type: none;
}
