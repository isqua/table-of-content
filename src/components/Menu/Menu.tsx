import { useLocation } from 'react-router-dom'

import { type TableOfContent } from '../../features/toc'
import { Section } from './Section/Section'
import { MenuProvider } from './Context/MenuProvider'

import styles from './Menu.module.css'

type MenuProps = {
    toc: TableOfContent
}

export function Menu({ toc }: MenuProps): JSX.Element {
    const location = useLocation()
    const currentUrl = location.pathname.replace(/^\//, '')

    return (
        <nav className={styles.menu}>
            <ul className={styles.list}>
                <MenuProvider toc={toc} url={currentUrl}>
                    <Section parentId='' level={0} />
                </MenuProvider>
            </ul>
        </nav>
    )
}
