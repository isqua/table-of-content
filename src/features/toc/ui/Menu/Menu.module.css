.menu {
    font-size: 13px;
    font-style: normal;
    font-weight: 400;
    line-height: 20px;

    --menu-item-block-padding: 8px;
    --menu-item-inline-start-padding: 26px;
    --menu-item-inline-end-padding: 22px;
    --menu-item-level-padding: 16px;

    /* Align chevrons with text even for multiline rows */
    --menu-chevron-vertical-correction: 2px;
    --menu-chevron-horizontal-correction: -20px;
}

.list {
    display: flex;
    flex-direction: column;
    padding: 0;
    margin: 0;
    list-style-type: none;
}
