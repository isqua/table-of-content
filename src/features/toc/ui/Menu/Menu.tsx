
import type { TableOfContent } from '../../types'
import { MenuProvider } from './Context/MenuProvider'
import { Filter } from './Filter/Filter'
import { List } from './List/List'

import styles from './Menu.module.css'

export type MenuProps = {
    toc: TableOfContent
    activeUrl: string
    isLoading?: boolean
}

export function Menu({ toc, activeUrl, isLoading }: MenuProps): JSX.Element {
    return (
        <nav className={styles.menu}>
            <MenuProvider toc={toc} url={activeUrl} isLoading={isLoading}>
                <Filter />
                <List />
            </MenuProvider>
        </nav>
    )
}
