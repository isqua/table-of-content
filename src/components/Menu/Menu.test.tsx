import { fireEvent, screen, waitFor } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import type { TableOfContent } from '../../features/toc'
import tocFlat from '../../test/fixtures/toc/flat.json'
import tocTwoLevels from '../../test/fixtures/toc/two-levels.json'
import { renderInApp } from '../../test'
import { Menu } from './Menu'

describe('components/Menu', () => {
    it('should render skeletons while TOC is loading', async () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/bar.html'

        act(() => {
            renderInApp(<Menu isLoading toc={toc} />, { url: currentUrl })
        })

        expect(await screen.findByRole('navigation')).toMatchSnapshot()
    })

    it('should build a menu and highlight current page', async () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/bar.html'

        renderInApp(<Menu toc={toc} />, { url: currentUrl })

        expect(await screen.findByRole('navigation')).toMatchSnapshot()
    })

    it('should build a two-levels menu and open all parents that contains current page', async () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/bar-install.html'

        renderInApp(<Menu toc={toc} />, { url: currentUrl })

        expect(await screen.findByRole('navigation')).toMatchSnapshot()
    })

    it('should close a submenu when clicking on a chevron', async () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/bar-install.html'

        renderInApp(<Menu toc={toc} />, { url: currentUrl })

        expect(await screen.findByRole('navigation')).toMatchSnapshot()

        fireEvent.click(screen.getByRole('button', { expanded: true }))

        expect(await screen.findByRole('navigation')).toMatchSnapshot()
    })

    it('should open a submenu when clicking on an item with children', async () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/foo.html'

        renderInApp(<Menu toc={toc} />, { url: currentUrl })

        fireEvent.click(screen.getByRole('link', { name: 'Bar' }))

        expect(await screen.findByRole('button', { expanded: true })).not.toBeNull()

        expect(screen.getByText('Bar: Install')).not.toBeNull()
    })
})
