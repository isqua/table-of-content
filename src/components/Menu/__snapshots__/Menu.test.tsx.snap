// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/Menu > should build a menu and highlight current page 1`] = `
<nav
  class="_menu_29d207"
>
  <ul
    class="_list_29d207"
  >
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _active_daf4a6"
        href="/bar.html"
      >
        <span
          class="_text_daf4a6"
        >
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/baz.html"
      >
        <span
          class="_text_daf4a6"
        >
          Baz
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/foo.html"
      >
        <span
          class="_text_daf4a6"
        >
          Foo
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`components/Menu > should build a two-levels menu and open all parents that contains current page 1`] = `
<nav
  class="_menu_29d207"
>
  <ul
    class="_list_29d207"
  >
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _parent_daf4a6"
        href="/bar.html"
      >
        <span
          class="_text_daf4a6"
        >
          <svg
            aria-expanded="true"
            class="_chevron_eafbf3 _toggle_daf4a6 _open_eafbf3"
            fill="none"
            height="16"
            role="button"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 13 11 8 6 3v9Z"
              fill="currentColor"
            />
          </svg>
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _active_daf4a6"
        href="/bar-install.html"
      >
        <span
          class="_text_daf4a6"
        >
          Bar: Install
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _parent_daf4a6"
        href="/bar-features.html"
      >
        <span
          class="_text_daf4a6"
        >
          Bar: Features
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/baz.html"
      >
        <span
          class="_text_daf4a6"
        >
          Baz
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/foo.html"
      >
        <span
          class="_text_daf4a6"
        >
          Foo
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`components/Menu > should close a submenu when clicking on a chevron 1`] = `
<nav
  class="_menu_29d207"
>
  <ul
    class="_list_29d207"
  >
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _parent_daf4a6"
        href="/bar.html"
      >
        <span
          class="_text_daf4a6"
        >
          <svg
            aria-expanded="true"
            class="_chevron_eafbf3 _toggle_daf4a6 _open_eafbf3"
            fill="none"
            height="16"
            role="button"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 13 11 8 6 3v9Z"
              fill="currentColor"
            />
          </svg>
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _active_daf4a6"
        href="/bar-install.html"
      >
        <span
          class="_text_daf4a6"
        >
          Bar: Install
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _parent_daf4a6"
        href="/bar-features.html"
      >
        <span
          class="_text_daf4a6"
        >
          Bar: Features
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/baz.html"
      >
        <span
          class="_text_daf4a6"
        >
          Baz
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/foo.html"
      >
        <span
          class="_text_daf4a6"
        >
          Foo
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`components/Menu > should close a submenu when clicking on a chevron 2`] = `
<nav
  class="_menu_29d207"
>
  <ul
    class="_list_29d207"
  >
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6 _parent_daf4a6"
        href="/bar.html"
      >
        <span
          class="_text_daf4a6"
        >
          <svg
            aria-expanded="false"
            class="_chevron_eafbf3 _toggle_daf4a6"
            fill="none"
            height="16"
            role="button"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 13 11 8 6 3v9Z"
              fill="currentColor"
            />
          </svg>
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/baz.html"
      >
        <span
          class="_text_daf4a6"
        >
          Baz
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <a
        class="_link_daf4a6"
        href="/foo.html"
      >
        <span
          class="_text_daf4a6"
        >
          Foo
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`components/Menu > should render skeletons while TOC is loading 1`] = `
<nav
  class="_menu_29d207"
>
  <ul
    class="_list_29d207"
  >
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <span
        class="_link_daf4a6"
      >
        <span
          class="_text_daf4a6 _skeleton_daf4a6"
        >
          <span
            class="_skeleton_2cc9eb _loader_daf4a6"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <span
        class="_link_daf4a6"
      >
        <span
          class="_text_daf4a6 _skeleton_daf4a6"
        >
          <span
            class="_skeleton_2cc9eb _loader_daf4a6"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="2"
      class="_item_daf4a6"
    >
      <span
        class="_link_daf4a6"
      >
        <span
          class="_text_daf4a6 _skeleton_daf4a6"
        >
          <span
            class="_skeleton_2cc9eb _loader_daf4a6"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <span
        class="_link_daf4a6"
      >
        <span
          class="_text_daf4a6 _skeleton_daf4a6"
        >
          <span
            class="_skeleton_2cc9eb _loader_daf4a6"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="1"
      class="_item_daf4a6"
    >
      <span
        class="_link_daf4a6"
      >
        <span
          class="_text_daf4a6 _skeleton_daf4a6"
        >
          <span
            class="_skeleton_2cc9eb _loader_daf4a6"
          />
        </span>
      </span>
    </li>
  </ul>
</nav>
`;
